src/constructUrl.js

```diff
@@ -113,49 +114,176 @@ var DEFAULT_OPTIONS = Object.freeze({
  * @return {String}             URL of image src transformed by Imgix
  */
 function constructUrl(src, longOptions) {
-  if (!src) {
-    return "";
-  }
-
-  const keys = Object.keys(longOptions);
-  const keysLength = keys.length;
-  let url = src + "?";
-  for (let i = 0; i < keysLength; i++) {
-    let key = keys[i];
-    let val = longOptions[key];
+  const params = compactParamKeys(longOptions);
+  const { client, pathComponents } = extractClientAndPathComponents(src);
+  return client.buildURL(pathComponents.join("/"), params);
+}
 
+function compactParamKeys(longOptions, width, height) {
+  const params = Object.keys(longOptions).reduce((acc, key) => {
     if (PARAM_EXPANSION[key]) {
-      key = PARAM_EXPANSION[key];
+      acc[PARAM_EXPANSION[key]] = longOptions[key];
     } else {
-      key = encodeURIComponent(key);
+      acc[key] = longOptions[key];
     }
+    return acc;
+  }, {});
 
-    if (key.substr(-2) === "64") {
-      val = Base64.encodeURI(val);
-    }
+  if (width) {
+    params["w"] = width;
+  }
 
-    url += key + "=" + encodeURIComponent(val) + "&";
+  if (height) {
+    params["h"] = height;
   }
+  return params;
+}
+
+function extractClientAndPathComponents(src) {
+  const [scheme, rest] = src.split("://");
+  const [domain, ...pathComponents] = rest.split("/");
+  let useHTTPS = scheme == "https";
+
+  const client = new ImgixClient({
+    domain: domain,
+    useHTTPS: useHTTPS,
+    includeLibraryParam: false,
+  });
 
-  return url.slice(0, -1);
+  return { client, pathComponents };
 }
 
 function buildURLPublic(src, imgixParams = {}, options = {}) {
   const { disableLibraryParam } = options;
 
   const [rawSrc, params] = extractQueryParams(src);
 
-  return constructUrl(
-    rawSrc,
-    Object.assign(
-      {},
-      params,
-      imgixParams,
-      disableLibraryParam ? {} : { ixlib: `react-${PACKAGE_VERSION}` }
-    )
-  );
+  return constructUrl(rawSrc, {
+    ...params,
+    ...imgixParams,
+    ...(disableLibraryParam ? {} : { ixlib: `react-${PACKAGE_VERSION}` }),
+  });
```

Comment:
Agreed! I can't walk this back... or really I wanted to benchmark things pre/post imgix/js-core integration; I can continue ahead with that work.. it might be more necessary than not at this point. I have seen both `Object.assign` and `Object spread` coming through perf tools, let me try to get some numbers.

Reply:
✅